1.train_models/log_reg_plate_1_cp_fs_data/nbconverted/train_model.py

```diff
@@ -0,0 +1,237 @@
+# ## Down-sample and stratify by well
+
+# In[ ]:
+
+
+smallest_gene = platedf[target_column].value_counts().min()
+platedata = pd.DataFrame()
+
+for gene in platedf[target_column].unique():
+    df = platedf.loc[platedf["Metadata_genotype"] == gene]
+    df_frac = smallest_gene / len(df)
+    stratwell = df.groupby(stratify_column, group_keys=False).apply(
+        lambda x: x.sample(frac=df_frac, random_state=rnd_val)
+    )
+    platedata = pd.concat([platedata, stratwell], axis="rows")
+
+
+# ## Stratified Train-test split
+
+# In[ ]:
+
+
+traindf, testdf = train_test_split(
+    platedata,
+    train_size=train_val_frac,
+    stratify=platedata[[target_column, stratify_column]],
+    shuffle=True,
+    random_state=rnd_val,
+)
+
+
+# ## Encode Labels
+
+# In[ ]:
+
+
+le = LabelEncoder()
+testdf["label"] = le.fit_transform(testdf[target_column].values)
+traindf["label"] = le.transform(traindf[target_column].values)
+
+
+# ## Remove unecessary columns
+
+# In[ ]:
+
+
+# Remove Metadata
+feat_col = [col for col in platedata.columns if "Metadata" not in col]
+
+# Keep the label column
+feat_col.append("label")
+
+traindf = traindf[feat_col]
+testdf = testdf[feat_col]
+
+
+# # Model Training
+
+# In[ ]:
+
+
+# Default number of splits
+num_splits = 5
+
+
+def kcross_val(model, X_data, y_data, splits=num_splits):
+    """
+    Parameters
+    ----------
+    model: sklearn model
+        The model to be trained and evaluated.
+    X_data : Pandas Dataframe
+        The preprocessed dataframe with features.
+    y_data : Pandas Dataframe
+        The preprocessed dataframe with labels.
+    splits: int
+        The number of splits for k cross validation
+
+    Returns
+    -------
+    best_model: Sklearn Model
+        The best model based on the validation accuracy
+    """
+
+    cv = KFold(n_splits=splits, random_state=rnd_val, shuffle=True)
+    max_acc = 0  # Keep track of the maximum accuracy
+
+    # iterates through index splits
+    for train_idx, val_idx in cv.split(X=X_data, y=y_data):
+
+        # Validation labels converted to ints
+        yval = y_data.iloc[val_idx].astype(np.int64)
+
+        # Training labels converted to ints
+        ytrain = y_data.iloc[train_idx].astype(np.int64)
+
+        # Validation features
+        Xval = X_data.iloc[val_idx]
+
+        # Training features
+        Xtrain = X_data.iloc[train_idx]
+
+        # Fitting the model
+        model.fit(Xtrain, ytrain)
+
+        # Scoring the model on the validation set
+        acc = model.score(Xval, yval)
+
+        # Find the model with the best validation accuracy
+        if max_acc < acc:
+            best_model = model
+
+    return best_model
+
+
+# In[1]:
+
+
+lr = LogisticRegression(max_iter=1000, solver="sag", random_state=rnd_val, n_jobs=-1)
+
+kcross_val(lr, X_data=traindf.drop("label", axis="columns"), y_data=traindf["label"])
+
+
+# ## Save Data
+
+# In[2]:
+
+
+dump(lr, data_path / "lr_model.joblib")
+dump(testdf, data_path / "testdf.joblib")
```

Comment:
what is `testdf`? If it was for initial prototyping/testing, please consider deleting

Reply:
testdf is for 2.evaluate_models

Reply:
I was considering grouping the evaluation and the training parts of the models together, but I decided to keep them separate to follow the structure of another analysis repo. Maybe grouping them would be better?

Reply:
I think separating is better :)

also, if `testdf` is used downstream, please use a more descriptive file name. Also, is it a pandas dataframe? If so, a compressed csv is probably better than saving it as a joblib file.

Reply:
Good point, I'll change that